### PR TITLE
Tim dropdown v1

### PR DESCRIPTION
**Pull request by timand1**
### Dropdown

First you need to install uuid and popperjs by
``` html
npm i
```

Dropdown is divided into 5 component - Dropdown Button, Dropdown Input, Dropdown Item, Dropdown Filter and Dropdown Divider.

Working - 
- Unit tests
- Storybook

> Storybook also a working dark theme.

### Dropdown Button

The Dropdown Button component have the Foam Button as a base for the dropdown. Width a ``slot`` in which elements can be inserted. Dropdown Item or Dropdown Filter for example.

``` html
  <foam-dropdown-button variant="primary" text="Links" icon="warning" > 
    <foam-dropdown-item />
    <foam-dropdown-item />
  </foam-dropdown-button> 
```

> Note - Dropdown Item is explained further down.

- Props - type - example 
- - variant - string - danger 
- - text - string - Dropdown Text
- - icon - string | null - warning
- - width - number | null - 250
- - align - string - start
- - disabled - boolean

> Note - Disabled is best used to pass a variable that can be changed in the parent.

### Dropdown Input

The dropdown Input is like Dropdown Button a component with a ``slot`` in which elements can be inserted. Dropdown Item or Dropdown Filter for example.

The Dropdown Input has an input filed that works like a search in which the user can write and the Dropdown Input component will filter the components in the ``slot`` dropdown.

``` html
  <foam-dropdown-input variant="primary" text="Links" icon="warning" > 
    <foam-dropdown-filter />
    <foam-dropdown-filter />
  </foam-dropdown-input> 
```

- Props - type - example 
- - text - string - Dropdown Text
 - - placeholder- string | null - Search items...

> Note - Dropdown Filter is explained further down.

### Dropdown Item

The Dropdown Item component is best used within a Dropdown Button or Dropdown Input component. 

It can be used as a button width a passed down ``click event`` by setting the type to ``button`` or by excluding passing a type value. Button is default.

It can also be used as a link - both within the application or external links.
type ``route`` links within and type ``link`` is external links.

The ``to`` property decides the to / href.
> With type ``route`` the ``to`` value is the endpoint in the application router
> With type ``link`` the ``to`` value is the full url link. 

- Props - type - example 
- - type - string - link
- - text - string | null - Dropdown Item
- - icon - string | null - warning
- - to - string - http://www.google.se / about-us
- - disabled - boolean
- - selected - boolean
- - secondaryText - string | null
- - width - number | null - 250
- - subdropdown - boolean

> Disabled -  best used to pass a variable that can be changed in the parent.
> Selected -  best used to pass a variable that can be changed in the parent.
> to - only used with type values ``link`` and ``route``.
> secondaryTest - only useable if the text property is greater than 0 in length.
> subdropdown -  only useable with type ``button``.
> width - alters the width of the sub dropdown.

If the property ``subdropdown`` is passed. The Dropdown Item will in essence work the same as Dropdown Button and Dropdown Input with a ``slot`` in which elements can be inserted.

### Dropdown Filter

The Dropdown Filter component is best used within a Dropdown Button or Dropdown Input component. 
Best used together with a passed down ``click event`` and a ``selected`` boolean variable.

- Props - type - example 
- - type - string - color
- - text - string - Blue
- - disabled - boolean
- - selected - boolean
- - color - string | null - #303A49

> Color - only used together with type ``color`` to set the color of the color filter element.
> Disabled -  best used to pass a variable that can be changed in the parent.
> Selected -  best used to pass a variable that can be changed in the parent.

### Dropdown Divider

A simple component that adds a line. Used between two Dropdown Items or Dropdown Filters for best effect.

``` html
  <foam-dropdown-input variant="primary" text="Links" icon="warning" > 
    <foam-dropdown-filter />
    <foam-dropdown-filter />
    <foam-dropdown-divider />
    <foam-dropdown-filter />
    <foam-dropdown-filter />
  </foam-dropdown-input> 
```

### CODE REVIEW

**Review by saradalsman — APPROVED** [review]
The tabtrap works great now! Well done 🚀

File: src/components/dropdown/dropdown-button/dropdown-button.vue [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    />
+    <div
+      id="dropdown"
+      ref="dropdown"
+      :class="dropdownId"
+      class="subtitle2"
+      :style="{ width: props.width ? props.width + 'px !important' : 'fit-content' }"
+      @keyup.escape="handleShowDropdown"
+    >
+      <slot />
```
Comment (adamfagerstrom) [inline src/components/dropdown/dropdown-button/dropdown-button.vue:110]: One thing you could think about in the future is to use named-slots. It's easier for the readability and also makes life easier if you decide to add more slots to a component.

**Review by adamfagerstrom — APPROVED** [review]
I have some few comments, nothing that's "important" to fix and should block merging. But could be nice to change.

For further dev. it would require us to actually use this component in a real-context. To test and see how it works when we pass down data e.tc. BUT that's outside of this scope.

What you have done here is beyond any requirements we may have communicated. So great work! Really!

File: src/composables/tabTrap.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  
+  if(
+    !currentElem.nextElementSibling?.hasAttribute('data-show') 
+    && !currentElem.nextElementSibling?.classList.contains('sub-dropdown__item')
+    && !currentElem.previousElementSibling?.classList.contains('sub-dropdown__item')
+    && !currentElem.className.includes('sub__item')
+    ) {    
+      elements.value = elements.value.filter(el => !el.classList.contains('sub-dropdown__item'))
+}
+
```
[the remark this answers is not in the document]
Comment (timand1) [reply src/composables/tabTrap.ts:24]: That is true (with the generic class) and also thought about it when I started working on the select component. I will see if I can find a better solution for it.

File: src/components/dropdown/dropdown-button/dropdown-button.vue (hunk 1 above)
Comment (timand1) [reply src/components/dropdown/dropdown-button/dropdown-button.vue:110]: Alright, even if you 'know' you'll only have one slot. Naming them should be the norm either way?

**Pull request merged**